interfaces: add access to pipewire socket

sergio-costas opened the pull request:
This MR adds a new interface, 'pipewire-server', that gives full access to the pipewire socket and lock file, and the pulseaudio pid file, located at

    /run/user/XXX/pipewire-0
    /run/user/XXX/pipewire-0.lock
    /run/user/XXX/pulse/pid

It allows not only to read and write to the socket, but also to create it. This is needed for the following things:

    - containerize the Pipewire/Wireplumber/Pipewire-pulse daemons
    - containerize xdg-desktop-portal
    - allow to share the screen in ubuntu-core-desktop
    - have pipewire support in ubuntu-core

This is NOT needed for normal use of pipewire, because xdg-desktop-portal is able to export it in a secure maner inside a container.

https://warthogs.atlassian.net/browse/UCD-102

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

Meulengracht reviewed (commented):
The added unit tests are failing

pedronis commented:
@jhenstridge said he needs to look a this first

tonyespy commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,151 @@
+
+const pipewireServerPermanentSlotAppArmor = `
+# When running Pipewire in system mode it will switch to the at
+# build time configured user/group on startup.
```
Please note that currently the only user/group that will work is the `snap_daemon` user (https://snapcraft.io/docs/system-usernames).

tonyespy commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,151 @@
+/sys/**/sound/** r,
+
+
+# Shared memory based communication with clients
```
This comment doesn't seem to apply to any of the AppArmor statements. In the `audio_playback` interface this comment prefaces the statement:

`/{run,dev}/shm/pulse-shm-* mrwk,
`

One other related note... PulseAudio when run in system mode automatically disables shared memory support. Does PipeWire actually implement shared memory support for the Pulse native protocol, and if so, does it also disable when PipeWire (or more specifically the PipeWire-pulse daemon) run in system mode?

sergio-costas replied:
But the idea is to run the code not in system mode, but as an user daemon, using

    apps:
      pipewire:
        command: snap/command-chain/pipewire
        daemon: simple
        passthrough:
          daemon-scope: user

sergio-costas replied:
You can see the snap here: https://github.com/sergio-costas/pipewire-snap

tonyespy reviewed (commented):
Hope you don't mind an unsolicited review...

One thing that I'm curious about is why a new interface is actually needed when `audio_record` and `audio_playback` provide sufficient slot snippets to allow the PulseAudio server to operate in strictly confined mode? Shouldn't we just extend those interfaces if additional AppArmor and/or Seccomp changes are needed in order to run PipeWire?

sergio-costas commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,151 @@
+)
+
+// The pipewire-server interface is the companion interface to the audio-record
+// interface. The design of this interface is based on the idea that the slot
```
My fault: I just copied the audio-playback file and modified it to ensure that it did work as I needed to allow to support a snapped version of pipewire (this is: having the pipewire server, wireplumber and pipewire-pulse running inside an snap). The idea is to use it in Core Desktop. Clearly I must review the comments...

sergio-costas commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,151 @@
+    deny-auto-connection: true
+`
+
+const pipewireServerConnectedPlugAppArmor = `
```
As commented, this isn't just for audio, but for the whole pipewire. The snapped applications get access through portals, it's true, but the portals are offered by xdg-desktop-portal application, which requires access to the pipewire socket. So if pipewire is running inside a snap, it must be able to create the socket in the "global" XDG_RUNTIME_DIR, where the external xdg-desktop-portal application expects to find it. Also, a future snapped xdg-desktop-portal should use this same interface to be able to access to the pipewire socket. Finally, Gnome Shell requires access to the pipewire socket to do screen sharing, and it expects to find it at the "global" XDG_RUNTIME_DIR folder (and also, a snapped Gnome Shell will require this interface to get access to the pipewire socket placed at the "global" XDG_RUNTIME_DIR).

sergio-costas replied:
Also, it's not a good idea to give access to the pipewire socket to any snapped application, because, as commented, they already can access it using portals, which is The Right Way(tm) :-D

sergio-costas commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,151 @@
+
+const pipewireServerConnectedPlugSecComp = `
+shmctl
+`
```
Mmm... good point... an application using this interface should also have the audio-* interfaces plugged.

sergio-costas commented:
@tonyespy Ok, I remodeled the patch, fixed the comments and removed everything that is common with audio-playback. I think that it's the best, because that way, any bugfix done in audio-playback won't have to be also done here. Can you review it again, please?

sergio-costas commented:
@jhenstridge Can you review this when you have some spare time, please?

tonyespy commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,111 @@
+owner /{,var/}run/user/[0-9]*/pipewire-0 rwk,
+owner /{,var/}run/user/[0-9]*/pipewire-0.lock rwk,
+owner /{,var/}run/user/[0-9]*/pulse/pid rwk,
+`
```
Like with the `*PlugAppArmor` above, this slot snippet duplicates** this line from `audio-playback`:

`owner /run/user/[0-9]*/ r,
`

And I don't think this line is needed (note, it's not present in `audio-playback` either):

`owner /{,var/}run/user/[0-9]*/pulse/pid rwk,`

** The line isn't an exact duplicate as the `audio-playback` interfaces prefixes the path with `/run`, whereas this interface uses the more generic `/{,var/}run`.

tonyespy commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,111 @@
+	return interfaces.StaticInfo{
+		Summary:              pipewireServerSummary,
+		ImplicitOnClassic:    true,
+		ImplicitOnCore:       true,
```
Why does this differ from `audio-playback`, I would think we'd only want this slot defined in the pipewire snap was installed?

sergio-costas replied:
I don't understand this... what do you mean?

tonyespy commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,111 @@
+        - app
+        - core
+    deny-connection:
+      on-classic: false
```
Does this imply `on-core: true`?

sergio-costas replied:
Well, honestly, I don't know... I'm quite new to snapd.

sergio-costas replied:
I just copied it from audio-playback.

sergio-costas commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,111 @@
+owner /{,var/}run/user/[0-9]*/ r,
+owner /{,var/}run/user/[0-9]*/pipewire-0 rwk,
+owner /{,var/}run/user/[0-9]*/pipewire-0.lock rwk,
+owner /{,var/}run/user/[0-9]*/pulse/pid rwk,
```
Not really about the pid one: the `audio-playback` interface only allows to read it; this one allows to create and write it. Also, it offers full access to the pipewire socket and the lock file because this interface is mainly oriented for snapping the pipewire server.

The other one... is possible, I'll check.

sergio-costas replied:
fixed.

sergio-costas commented on interfaces/builtin/pipewire_server.go:
```diff
@@ -0,0 +1,111 @@
+owner /{,var/}run/user/[0-9]*/pipewire-0.lock rwk,
+owner /{,var/}run/user/[0-9]*/pulse/pid rwk,
+`
+
```
I presume that, since snapd isn't the one that provides this, it's not needed.

sergio-costas replied:
I'll test it.

sergio-costas replied:
It works without it. Removed.

sergio-costas commented:
@jhenstridge Found that there is at least one program, g4music, that seems to access pipewire-0 socket directly... 

https://gitlab.gnome.org/neithern/g4music/-/blame/master/pkgs/flatpak/com.github.neithern.g4music.json#L17

Something like this, but only for clients, would be useful... What do you think?

soumyaDghosh commented:
> @jhenstridge Found that there is at least one program, g4music, that seems to access pipewire-0 socket directly...
> 
> https://gitlab.gnome.org/neithern/g4music/-/blame/master/pkgs/flatpak/com.github.neithern.g4music.json#L17
> 
> Something like this, but only for clients, would be useful... What do you think?

To add this, [clapper](https://github.com/Rafostar/clapper) and [Cavalier](https://github.com/NickvisionApps/Cavalier) is also lacking and needs this

pedronis commented:
I agree with the points @jhenstridge raised. I think this should be closed until we have a more detailed plan about if/how to grant mediated access to pipewire directly
